README
NAME
    Declare::Constraints::Simple - Declarative Validation of Data Structures

SYNOPSIS
      use Declare::Constraints::Simple-All;

      my $profile = IsHashRef(
                        -keys   => HasLength,
                        -values => IsArrayRef( IsObject ));

      my $result1 = $profile->(undef);
      print $result1->message, "\n";    # 'Not a HashRef'

      my $result2 = $profile->({foo => [23]});

      print $result2->message, "\n";    # 'Not an Object'

      print $result2->path, "\n";       
                        # 'IsHashRef[val foo].IsArrayRef[0].IsObject'

DESCRIPTION
    The main purpose of this module is to provide an easy way to build a
    profile to validate a data structure. It does this by giving you a set
    of declarative keywords in the importing namespace.

USAGE
    This is just a brief intro. For details read the documents mentioned in
    "SEE ALSO".

  Constraint Import
      use Declare::Constraints::Simple-All;

    The above command imports all constraint generators in the library into
    the current namespace. If you want only a selection, use "only":

      use Declare::Constraints::Simple
          Only => qw(IsInt Matches And);

    You can find all constraints (and constraint-like generators, like
    operators. In fact, "And" above is an operator. They're both implemented
    equally, so the distinction is a merely philosophical one) documented in
    the Declare::Constraints::Simple::Library pod. In that document you will
    also find the exact parameters for their usage, so this here is just a
    brief Intro and not a coverage of all possibilities.

  Building a Profile
    You can use these constraints by building a tree that describes what
    data structure you expect. Every constraint can be used as
    sub-constraint, as parent, if it accepts other constraints, or
    stand-alone. If you'd just say

      my $check = IsInt;
      print "yes!\n" if $check->(23);

    it will work too. This also allows predefining tree segments, and
    nesting them:

      my $id_to_objects = IsArrayRef(IsObject);

    Here $id_to_objects would give it's OK on an array reference containing
    a list of objects. But what if we now decide that we actually want a
    hashref containing two lists of objects? Behold:

      my $object_lists = 
        IsHashRef( HasAllKeys( qw(good bad) ),
                   OnHashKeys( good => $id_to_objects,
                               bad  => $id_to_objects ));

    As you can see, constraints like "IsArrayRef" and "IsHashRef" allow you
    to apply constraints to their keys and values. With this, you can step
    down in the data structure.

  Applying a Profile to a Data Structure
    Constraints return just code references that can be applied to one value
    (and only one value) like this:

      my $result = $object_lists->($value);

    After this call $result contains a Declare::Constraints::Simple::Result
    object. The first think one wants to know is if the validation
    succeeded:

      if ($result->is_valid) { ... }

    This is pretty straight forward. To shorten things the result object
    also overloads it's "bool"ean context. This means you can alternatively
    just say

      if ($result) { ... }

    However, if the result indicates a invalid data structure, we have a few
    options to find out what went wrong. There's a human parsable message in
    the "message" accessor. You can override these by forcing it to a
    message in a subtree with the "Message" declaration. The "stack"
    contains the name of the chain of constraints up to the point of
    failure.

    You can use the "path" accessor for a joined string path representing
    the stack.

  Creating your own Libraries
    You can declare a package as a library with

      use Declare::Constraints::Simple-Library;

    which will install the base class and helper methods to define
    constraints. For a complete list read the documentation in
    Declare::Constraints::Simple::Library::Base. You can use other libraries
    as base classes to include their constraints in your export
    possibilities. This means that with a package setup like

      package MyLibrary;
      use warnings;
      use strict;

      use Declare::Constraints::Simple-Library;
      use base 'Declare::Constraints::Simple::Library';

      constraint 'MyConstraint',
        sub { return _result(($_[0] >= 12), 'Value too small') };

      1;

    you can do

      use MyLibrary-All;

    and have all constraints, from the default library and yours from above,
    installed into your requesting namespace. You can override a constraint
    just by redeclaring it in a subclass.

  Scoping
    Sometimes you want to validate parts of a data structure depending on
    another part of it. As of version 2.0 you can declare scopes and store
    results in them. Here is a complete example:

      my $constraint =
        Scope('foo',
          And(
            HasAllKeys( qw(cmd data) ),
            OnHashKeys( 
              cmd => Or( SetResult('foo', 'cmd_a',
                           IsEq('FOO_A')),
                         SetResult('foo', 'cmd_b',
                           IsEq('FOO_B')) ),
              data => Or( And( IsValid('foo', 'cmd_a'),
                               IsArrayRef( IsInt )),
                          And( IsValid('foo', 'cmd_b'),
                               IsRegex )) )));

    This profile would accept a hash references with the keys "cmd" and
    "data". If "cmd" is set to "FOO_A", then "data" has to be an array ref
    of integers. But if "cmd" is set to "FOO_B", a regular expression is
    expected.

SEE ALSO
    Declare::Constraints::Simple::Library,
    Declare::Constraints::Simple::Result,
    Declare::Constraints::Simple::Base, Module::Install

REQUIRES
    Carp::Clan, aliased, Class::Inspector, Scalar::Util, overload and
    Test::More (for build).

TODO
    *   More tests of course!

    *   An "OnArrayElements" constraint. Like "OnHashKeys" but for array
        references.

    *   Examples.

    *   A list of questions that might come up, together with their answers.

    *   A "Custom" constraint that takes a code reference.

    *   Create stack objects that stringify to the current form, but can
        hold more data.

    *   Give the "Message" constraint the ability to get the generated
        constraint inserted in the message. A possibility would be to
        replace __Value__ and __Message__. It might also accept code
        references, which return strings.

    *   Allow the "IsCodeRef" constraint to accept further constraints. One
        might like to check, for example, the refaddr of a closure.

    *   A "Captures" constraint that takes a regex and can apply other
        constraints to the matches.

    *   ???

    *   Profit.

INSTALLATION
      perl Makefile.PL
      make
      make test
      make install

    For details read Module::Install.

AUTHOR
    Robert 'phaylon' Sedlacek "<phaylon@dunkelheit.at>"

LICENSE AND COPYRIGHT
    This module is free software, you can redistribute it and/or modify it
    under the same terms as perl itself.



Changes

[0.02] Sun Sep 10 23:22:55 CEST 2006
    - Added IsEq scalar constraint.
    - Added Scope, SetResult and IsValid.
    - Fixed error in Or. Correct results are returned now.
    - Fixed error in OnHashKeys. Hash key constraints are now processed
      in the order they are defined.
    - Custom libraries are now possible.
    - Much better constraint POD.
    - Interior of Message constraint is now in the _with_message sub.
    - Introduced use D:C:S-Library import for libraries.
    - Split of the Library into Sub-Libraries. The Library package itself
      is now just a bundle of all shipped libraries. It's POD serves as
      docmap for the constraints.
    - Corrected minor POD typos.
    - Stack info now escapes [ and ].
    - General TODO updates.
    - Split of the USAGE section into subsections.
    - Added a Custom constraint to the TODO list.
    - Forgot to add date to Changes file. /me slaps self.

[0.01] Sat Sep  9 18:22:08 CEST 2006
    - Initial Release.




lib/Declare/Constraints/Simple/Library.pm
=head1 NAME

Declare::Constraints::Simple::Library - Constraint Library Bundle

=cut

package Declare::Constraints::Simple::Library;
use warnings;
use strict;

use base qw(
    Declare::Constraints::Simple::Library::General
    Declare::Constraints::Simple::Library::Scalar
    Declare::Constraints::Simple::Library::Numerical
    Declare::Constraints::Simple::Library::OO
    Declare::Constraints::Simple::Library::Referencial
    Declare::Constraints::Simple::Library::Hash
    Declare::Constraints::Simple::Library::Array
    Declare::Constraints::Simple::Library::Operators
);

=head1 DESCRIPTION

This module functions as bundle of all default libraries, and as map
and/or reference of said ones.

=head1 LIBRARIES

=over

=item L<Declare::Constraints::Simple::Library::General>

General constraints and constraint-like elements that affect the whole
framework.

Provides: C<Message>, C<Scope>, C<SetResult>, C<IsValid>

=item L<Declare::Constraints::Simple::Library::Scalar>

Constraints for scalar value validation.

Provides: C<Matches>, C<IsDefined>, C<HasLength>, C<IsOneOf>, C<IsTrue>,
C<IsEq>

=item L<Declare::Constraints::Simple::Library::Numerical>

These validate values by their numerical properties.

Provides: C<IsNumber>, C<IsInt>

=item L<Declare::Constraints::Simple::Library::OO>

For validation of values in an object oriented manner.

Provides: C<IsA>, C<IsClass>, C<IsObject>, C<HasMethods>

=item L<Declare::Constraints::Simple::Library::Referencial>

These can validate properties by their reference types.

Provides: C<IsRefType>, C<IsScalarRef>, C<IsArrayRef>, C<IsHashRef>,
C<IsCodeRef>, C<IsRegex>

=item L<Declare::Constraints::Simple::Library::Array>

These constraints deal with array references and their contents.

Provides: C<HasArraySize>

=item L<Declare::Constraints::Simple::Library::Hash>

All constraints appliable to hash references as well as their keys and
values.

Provides: C<HasAllKeys>, C<OnHashKeys>

=item Declare::Constraints::Simple::Library::Operators

Operators can be used in any place a constraint can be used, as
their implementations are similar.

Provides: C<And>, C<Or>, C<XOr>, C<Not>

=back

=head1 SEE ALSO

L<Declare::Constraints::Simple>

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;



lib/Declare/Constraints/Simple/Library/Array.pm
=head1 NAME

Declare::Constraints::Simple::Library::Array - Array Constraints

=cut

package Declare::Constraints::Simple::Library::Array;
use warnings;
use strict;

use Declare::Constraints::Simple-Library;

=head1 SYNOPSIS

  # accept a list of pairs
  my $pairs_validation = IsArrayRef( HasArraySize(2,2) );

=head1 DESCRIPTION

This module contains all constraints that can be applied to array
references.

=head1 CONSTRAINTS

=head2 HasArraySize([$min, [$max]])

With C<$min> defaulting to 1. So a specification of

  my $profile = HasArraySize;

checks for at least one value. To force an exact size of the array,
specify the same values for both:

  my $profile = HasArraySize(3, 3);

=cut

constraint 'HasArraySize',
    sub {
        my ($min, $max) = @_;
        $min = 1 unless defined $min;
        return sub {
            return _false('Undefined Value') unless defined $_[0];
            return _false('Not an ArrayRef') 
                unless ref($_[0]) eq 'ARRAY';
            return _false("Less than $min Array elements")
                unless scalar(@{$_[0]}) >= $min;
            return _true 
                unless $max;
            return _false("More than $max Array elements")
                unless scalar(@{$_[0]}) <= $max;
            return _true;
        };
    };

=head1 SEE ALSO

L<Declare::Constraints::Simple>, L<Declare::Constraints::Simple::Library>

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;


lib/Declare/Constraints/Simple/Library/General.pm
=head1 NAME

Declare::Constraints::Simple::Library::General - General Constraints

=cut

package Declare::Constraints::Simple::Library::General;
use warnings;
use strict;

use Declare::Constraints::Simple-Library;

use Carp::Clan qw(^Declare::Constraints::Simple);

=head1 SYNOPSIS

  # custom error messages
  my $constraint = 
    And( Message( 'You need to specify a Value', IsDefined ),
         Message( 'The specified Value is not an Int', IsInt ));

=head1 DESCRIPTION

This library is meant to contain those constraints and constraint-like
elements that apply generally to the whole framework.

=head1 CONSTRAINTS

=head2 Message($message, $constraint)

Overrides the C<message> set on the result object for failures in
C<$constraint>. For example:

  my $message = 'How hard is it to give me a number?';
  my $constraint = Message($message, IsNumber);

  my $result = $constraint->('duh...');
  print $result->message, "\n";

The C<Message> constraint overrides the error message returned by it's
whole subtree, however, the C<Message> specification nearest to the point
of failure will win. So while this

  my $constraint = Message( 'Foo',
                            IsArrayRef( Message( 'Bar', IsInt )));

  my $result = $constraint->(['I am not an Integer']);
  print $result->message;

will print C<Bar>, this

  my $result = $constraint->('I\'m not even an ArrayRef');
  print $result->message;

will output C<Foo>.

=cut

constraint 'Message',
    sub {
        my ($msg, $c) = @_;
        return sub {
            return _with_message($msg, $c, @_);
        };
    };

=head2 Scope($name, $constraint)

Executes the passed C<$constraint> in a newly generated scope named
C<$name>.

=cut

constraint 'Scope',
    sub {
        my ($scope_name, $constraint) = @_;
        return sub {
            return _with_scope($scope_name, $constraint, @_);
        };
    };

=head2 SetResult($scope, $name, $constraint)

Stores the result ov an evaluation of C<$constraint> in C<$scope> under
C<$name>.

=cut

constraint 'SetResult',
    sub {
        my ($scope, $name, $constraint) = @_;
        return sub {
            my $result = $constraint->(@_);
            _set_result($scope, $name, $result);
            return $result;
        };
    };

=head2 IsValid($scope, $name)

Returns a true result if the result C<$name>, which has to have been stored
previously in the scope named C<$scope>, was valid.

=cut

constraint 'IsValid',
    sub {
        my ($scope, $name) = @_;
        return sub {
            _info("$scope:$name");
            return _false unless _has_result($scope, $name);
            my $result = _get_result($scope, $name);
            return _result($result, 
                "Value '$name' in scope '$scope' is invalid");
        };
    };

=head1 SEE ALSO

L<Declare::Constraints::Simple>, L<Declare::Constraints::Simple::Library>

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;


lib/Declare/Constraints/Simple/Library/Operators.pm
=head1 NAME

Declare::Constraints::Simple::Library::Operators - Operators

=cut

package Declare::Constraints::Simple::Library::Operators;
use warnings;
use strict;

use Declare::Constraints::Simple-Library;

use Carp::Clan qw(^Declare::Constraints::Simple);

=head1 SYNOPSIS

  # all hast to be valid
  my $and_constraint = And( IsInt,
                            Matches(qr/0$/) );

  # at least one has to be valid
  my $or_constraint = Or( IsInt, HasLength );

  # only one can be valid
  my $xor_constraint = XOr( IsClass, IsObject );

  # reverse validity
  my $not_an_integer = Not( IsInt );

=head1 DESCRIPTION

This module contains the frameworks operators. These constraint like
elements act on the validity of passed constraints.

=head1 OPERATORS

=head2 And(@constraints)

Is true if all passed C<@constraints> are true on the value. Returns
the result of the first failing constraint.

=cut

constraint 'And',
    sub {
        my @vc = @_;
        return sub {
            for (@vc) {
                my $r = $_->($_[0]);
                return $r unless $r->is_valid;
            }
            return _true;
        };
    };

=head2 Or(@constraints)

Is true if at least one of the passed C<@contraints> is true. Returns the
last failing constraint's result if false.

=cut

constraint 'Or',
    sub {
        my @vc = @_;
        return sub {
            my $last_r;
            for (0 .. $#vc) {
                my $v = $vc[$_];
                my $r = $v->($_[0]);
                return _true if $r->is_valid;
                return $r if $_ == $#vc;
            }
            return _false('No constraints');
        };
    };

=head2 XOr(@constraints)

Valid only if a single one of the passed C<@constraints> is valid. Returns
the last failing constraint's result if false.

=cut

constraint 'XOr',
    sub {
        my @vc = @_;
        return sub {
            my $m = 0;
            for (@vc) {
                my $r = $_->($_[0]);
                $m++ if $r->is_valid;
            }
            return _result(($m == 1), sprintf 'Got %d true returns', $m);
        };
    };

=head2 Not($constraint)

This is valid if the passed C<$constraint> is false. The main purpose
of this operator is to allow the easy reversion of a constraint's 
trueness.

=cut

constraint 'Not',
    sub {
        my ($c) = @_;
        croak '\'Not\' only accepts only a constraint as argument'
            if defined $c and not ref($c) eq 'CODE';
        return sub {
            return _true unless $c;
            my $r = $c->($_[0]);
            return _false('Constraint returned true') if $r->is_valid;
            return _true;
        };
    };

=head1 SEE ALSO

L<Declare::Constraints::Simple>, L<Declare::Constraints::Simple::Library>

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;


lib/Declare/Constraints/Simple/Library/Exportable.pm
=head1 NAME

Declare::Constraints::Simple::Library::Exportable - Export Facilities

=cut

package Declare::Constraints::Simple::Library::Exportable;
use warnings;
use strict;

use Carp::Clan qw(^Declare::Constraints::Simple);
use Class::Inspector;

use aliased 'Declare::Constraints::Simple::Library::Base' => 'LibraryBase';
sub Library () { 'Declare::Constraints::Simple::Library' }

=head1 DESCRIPTION

This contains the constraint export logic of the module.

=head1 METHODS

=head2 import($flag, @args)

  use ExportableModule->All;

  # or
  use ExportableModule-Only => qw(Constraint1 ...);

  # or
  use ExportableModule-Library;

Exports the constraints to the calling namespace. This includes all
libraries in L<Declare::Constraints::Simple::Library>, that package itself
(providing all default constraints) or L<Declare::Constraints::Simple>
itself as a shortcut.

Possible flags are

=over

=item All

Imports all constraints registered in the class and its base classes.

=item Only

  use Declare::Constraints::Simple::Library::Scalar-Only => 'HasLength';

The above line would only import the C<HasLength> constraints from the
C<Scalar> default library. Note however, that you could also just have
said

  use Declare::Constraints::Simple-Only => 'HasLength';

as both C<::Simple> and C<::Simple::Library> work on all default
libraries.

=item Library

You can use this to define your own constraint library. For more
information, see L<Declare::Constraints::Simple::Library::Base>.

=back

=cut

sub import {
    my ($class, $flag, @args) = @_;
    return unless $flag;

    my $handle_map = $class->_build_handle_map;
    my $target = scalar(caller);
    
    if ($flag =~ /^-?all$/i) {
        $class->_export_all($target, $handle_map);
    }
    elsif ($flag =~ /^-?only$/i) {
        $class->_export_these($target, $handle_map, @args);
    }
    elsif ($flag =~ /^-?library$/i) {
        LibraryBase->install_into($target);
    }

    1;
}

=head2 _build_handle_map()

Internal method to build constraint-to-class mappings.

=cut

sub _build_handle_map {
    my ($class) = @_;

    if ($class eq 'Declare::Constraints::Simple') {
        $class = Library;
    }

    if ($class eq Library) {
        unless (Class::Inspector->loaded(Library)) {
            require Class::Inspector->filename(Library);
        }
    }

    my (%seen, %handle_map, @walk, %walked);
    @walk = do {
        no strict 'refs'; 
        ($class, @{$class . '::ISA'});
    };

    while (my $w = shift @walk) {

        next if $walked{$w};
        $walked{$w} = 1;

        if ($w->can('fetch_constraint_declarations')) {
            my @decl = $w->fetch_constraint_declarations;
            for my $d (@decl) {
                next if exists $seen{$d};
                $seen{$d} = 1;
                $handle_map{$d} = $w;
            }
        }

        push @walk,
            grep { not exists $walked{$_} }
              do { no strict 'refs' ; @{$w . '::ISA'} };
    }

    return \%handle_map;
}

=head2 _export_all($target, $handle_map)

Internal method. Exports all handles in C<$handle_map> into the C<$target> 
namespace.

=cut

sub _export_all {
    my ($class, $target, $handle_map) = @_;
    return $class->_export_these($target, $handle_map, keys %$handle_map);
}

=head2 _export_these($target, $handle_map, @constraints)

Internal method. Exports all C<@constraints> from C<$handle_map> into the
C<$target> namespace.

=cut

sub _export_these {
    my ($class, $target, $handle_map, @decl) = @_;

    for my $d (@decl) {
        my $gen = $handle_map->{$d}->fetch_constraint_generator($d);

        croak sprintf 
            'Constraint Generator for $s in %s did not return a closure',
            $d, $handle_map->{$d}
            unless ref($gen) eq 'CODE';

        {   no strict 'refs';
            *{$target . '::' . $d} = $gen;
        }
    }
}

=head1 SEE ALSO

L<Declare::Constraints::Simple>, L<Declare::Constraints::Simple::Library>,
L<Declare::Constraints::Simple::Library::Base>

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;


lib/Declare/Constraints/Simple.pm
=head1 NAME

Declare::Constraints::Simple - Declarative Validation of Data Structures

=cut

package Declare::Constraints::Simple;
use warnings;
use strict;

use base 'Declare::Constraints::Simple::Library::Exportable';

our $VERSION = 0.02;

=head1 SYNOPSIS

  use Declare::Constraints::Simple-All;

  my $profile = IsHashRef(
                    -keys   => HasLength,
                    -values => IsArrayRef( IsObject ));

  my $result1 = $profile->(undef);
  print $result1->message, "\n";    # 'Not a HashRef'

  my $result2 = $profile->({foo => [23]});

  print $result2->message, "\n";    # 'Not an Object'

  print $result2->path, "\n";       
                    # 'IsHashRef[val foo].IsArrayRef[0].IsObject'

=head1 DESCRIPTION

The main purpose of this module is to provide an easy way to build a
profile to validate a data structure. It does this by giving you a set of
declarative keywords in the importing namespace.

=head1 USAGE

This is just a brief intro. For details read the documents mentioned in
L<SEE ALSO>.

=head2 Constraint Import

  use Declare::Constraints::Simple-All;

The above command imports all constraint generators in the library into
the current namespace. If you want only a selection, use C<only>:

  use Declare::Constraints::Simple
      Only => qw(IsInt Matches And);

You can find all constraints (and constraint-like generators, like
operators. In fact, C<And> above is an operator. They're both implemented
equally, so the distinction is a merely philosophical one) documented in
the L<Declare::Constraints::Simple::Library> pod. In that document you
will also find the exact parameters for their usage, so this here is just
a brief Intro and not a coverage of all possibilities.

=head2 Building a Profile

You can use these constraints by building a tree that describes what data
structure you expect. Every constraint can be used as sub-constraint, as
parent, if it accepts other constraints, or stand-alone. If you'd just 
say

  my $check = IsInt;
  print "yes!\n" if $check->(23);

it will work too. This also allows predefining tree segments, and nesting
them:

  my $id_to_objects = IsArrayRef(IsObject);

Here C<$id_to_objects> would give it's OK on an array reference 
containing a list of objects. But what if we now decide that we actually 
want a hashref containing two lists of objects? Behold:

  my $object_lists = 
    IsHashRef( HasAllKeys( qw(good bad) ),
               OnHashKeys( good => $id_to_objects,
                           bad  => $id_to_objects ));

As you can see, constraints like C<IsArrayRef> and C<IsHashRef> allow you
to apply constraints to their keys and values. With this, you can step
down in the data structure.

=head2 Applying a Profile to a Data Structure

Constraints return just code references that can be applied to one value
(and only one value) like this:

  my $result = $object_lists->($value);

After this call C<$result> contains a
L<Declare::Constraints::Simple::Result> object. The first think one wants
to know is if the validation succeeded:

  if ($result->is_valid) { ... }

This is pretty straight forward. To shorten things the result object also
L<overload>s it's C<bool>ean context. This means you can alternatively
just say

  if ($result) { ... }

However, if the result indicates a invalid data structure, we have a few
options to find out what went wrong. There's a human parsable message in
the C<message> accessor. You can override these by forcing it to a 
message in a subtree with the C<Message> declaration. The C<stack> 
contains the name of the chain of constraints up to the point of failure.

You can use the C<path> accessor for a joined string path representing 
the stack.

=head2 Creating your own Libraries

You can declare a package as a library with

  use Declare::Constraints::Simple-Library;

which will install the base class and helper methods to define
constraints. For a complete list read the documentation in
L<Declare::Constraints::Simple::Library::Base>. You can use other
libraries as base classes to include their constraints in your export
possibilities. This means that with a package setup like

  package MyLibrary;
  use warnings;
  use strict;

  use Declare::Constraints::Simple-Library;
  use base 'Declare::Constraints::Simple::Library';

  constraint 'MyConstraint',
    sub { return _result(($_[0] >= 12), 'Value too small') };

  1;

you can do

  use MyLibrary-All;

and have all constraints, from the default library and yours from above,
installed into your requesting namespace. You can override a constraint
just by redeclaring it in a subclass.

=head2 Scoping

Sometimes you want to validate parts of a data structure depending on
another part of it. As of version 2.0 you can declare scopes and store
results in them. Here is a complete example:

  my $constraint =
    Scope('foo',
      And(
        HasAllKeys( qw(cmd data) ),
        OnHashKeys( 
          cmd => Or( SetResult('foo', 'cmd_a',
                       IsEq('FOO_A')),
                     SetResult('foo', 'cmd_b',
                       IsEq('FOO_B')) ),
          data => Or( And( IsValid('foo', 'cmd_a'),
                           IsArrayRef( IsInt )),
                      And( IsValid('foo', 'cmd_b'),
                           IsRegex )) )));

This profile would accept a hash references with the keys C<cmd> and
C<data>. If C<cmd> is set to C<FOO_A>, then C<data> has to be an array 
ref of integers. But if C<cmd> is set to C<FOO_B>, a regular expression 
is expected.

=head1 SEE ALSO

L<Declare::Constraints::Simple::Library>, 
L<Declare::Constraints::Simple::Result>,
L<Declare::Constraints::Simple::Base>,
L<Module::Install>

=head1 REQUIRES

L<Carp::Clan>, L<aliased>, L<Class::Inspector>, L<Scalar::Util>,
L<overload> and L<Test::More> (for build).

=head1 TODO

=over

=item *

More tests of course!

=item *

An C<OnArrayElements> constraint. Like C<OnHashKeys> but for array
references.

=item *

Examples.

=item *

A list of questions that might come up, together with their answers.

=item *

A C<Custom> constraint that takes a code reference.

=item *

Create stack objects that stringify to the current form, but can hold
more data.

=item *

Give the C<Message> constraint the ability to get the generated 
constraint inserted in the message. A possibility would be to replace 
__Value__ and __Message__. It might also accept code references, which 
return strings.

=item *

Allow the C<IsCodeRef> constraint to accept further constraints. One 
might like to check, for example, the refaddr of a closure.

=item *

A C<Captures> constraint that takes a regex and can apply other
constraints to the matches.

=item *

???

=item *

Profit.

=back

=head1 INSTALLATION

  perl Makefile.PL
  make
  make test
  make install

For details read L<Module::Install>.

=head1 AUTHOR

Robert 'phaylon' Sedlacek C<E<lt>phaylon@dunkelheit.atE<gt>>

=head1 LICENSE AND COPYRIGHT

This module is free software, you can redistribute it and/or modify it 
under the same terms as perl itself.

=cut

1;
